chain/chain/src/store.rs

```diff
@@ -1734,14 +1735,47 @@ impl<'a> ChainStoreUpdate<'a> {
             .insert((height, shard_id), chunk_hash);
     }
 
-    pub fn clear_old_data_on_height(&mut self, height: BlockHeight) -> Result<(), Error> {
+    pub fn clear_old_data_on_height(
+        &mut self,
+        trie: Arc<Trie>,
+        height: BlockHeight,
+    ) -> Result<(), Error> {
         let mut store_update = self.store().store_update();
         let blocks_current_height = match self.get_all_block_hashes_by_height(height) {
             Ok(blocks_current_height) => {
                 blocks_current_height.values().flatten().cloned().collect()
             }
             _ => vec![],
         };
+
+        // 1. Trie cleaning
```

Comment:
Unclear what the contract of this function is:
- If the block is on a fork, we delete the state that's the result of applying this block
- If the block is on main chain and old, we delete the state that's **before** applying this block

we need to be careful here